File: lab/lab7/Dictionary.java
// Dictionary.java
// Juan Andreas
// jandreas
// lab7
// Linked List implementation of the Dictionary ADT

import java.io.*;
import java.util.Scanner;
import java.util.Arrays;

public class Dictionary implements DictionaryInterface{
	
   //--------------------------------------------------------------------------
   // Initialize Node 
   private class Node{
      Pair item;
	  Node left;
	  Node right;

      Node(Pair x){
		 item = x;
         left = null;
		 right = null;
      }
   }
   
   // private inner Pair class
   private class Pair{
	   String key;
	   String value;
	   
	   Pair(String x, String y){
		   key = x;
		   value = y;
	   }
   }
   
   // Fields
   private Node root;
   private int numPairs;

   // IntegerList()
   // constructor
   public Dictionary(){
      root = null;
      numPairs = 0;
   }
   
   //findKey()
   // returns the Node containing key k in the subtree rooted at R, or returns
   // null if no such Node exists
   private Node findKey(Node R, String k){
	   if(R==null || R.item.key.equals(k)){
		   return R;
	   }
	   if(R.item.key.compareTo(k)>0)
		   return findKey(R.left, k);
	   else
		   return findKey(R.right, k);
   }
   
   Node findParent(Node N, Node R){
	   Node P = null;
	   if(N!=R){
		   P = R;
		   while(P.left!=N && P.right!=N){
			   if(N.item.key.compareTo(P.item.key)<0)
				   P = P.left;
			   else
				   P = P.right;
		   }
	   }
	   return P;
   }
   
   //findLeftmost()
   // returns the leftmost Node in the subtree rooted at R, or null if R is null
   Node findLeftmost(Node R){
	   Node L = R;
	   if(L!=null) for( ; L.left!=null; L=L.left);
	   return L;
   }
   
   // printInOrder()
   // prints the (key, value) pairs belonging to the subtree rooted at R in order
   // of increasing keys to file pointed to by out
   void printInOrder(Node R){
	   if(R!=null){
		   printInOrder(R.left);
		   System.out.println(R.item.key+" "+R.item.value);
		   printInOrder(R.right);
	   }
   }
   
   // deleteAll()
   // deletes all Nodes in the subtree rooted at N
   void deleteAll(Node N){
	   if(N!=null){
		   deleteAll(N.left);
		   deleteAll(N.right);
	   }
   }
   
   // public functions ----------------------------------------------
   
   // isEmpty()
   public boolean isEmpty(){
	   return(numPairs==0);
   }
 
   // size()
   // pre: none
   // post: returns the number of elements in this Dictionary
   public int size() {
     return numPairs;
   }
   
   // lookup()
   // pre: none
   // post: returns value to associated key, or null if key doesn't exists
   public String lookup(String k){
	   Node N;
	   N = findKey(root, k);
	   return( N==null ? null : N.item.value);
   }
   
   // insert()
   // inserts new (key,value) pair into this Dictionary
   // pre: lookup(key)==null
   // post: !isEmpty(), size() is increased by one
   public void insert(String k, String v) throws DuplicateKeyException{
	   Node N, A, B;
	   if(findKey(root, k)!=null){
		   throw new DuplicateKeyException("Dictionary Error: cannot insert duplicate keys");
	   }
	   N = new Node(new Pair(k, v));
	   B = null;
	   A = root;
	   while(A!=null){
		   B = A;
		   if(A.item.key.compareTo(k)>0) A = A.left;
		   else A = A.right;
	   }
	   if(B==null) root = N;
	   else if(B.item.key.compareTo(k)>0) B.left = N;
	   else B.right = N;
	   numPairs++;
   }
   
   // delete()
   // deletes pair with the given key
   // pre: lookup(D, key)!=null
   public void delete(String k) throws KeyNotFoundException{
	   Node N, P, S;
	   if(findKey(root, k)==null){
		   throw new KeyNotFoundException("Dictionary Error: cannot delete non-existent key");
	   }
	   N = findKey(root, k);
	   if(N==null){
		   throw new KeyNotFoundException("Dictionary Error: cannot delete non-existent key");
	   }
	   if(N.left==null && N.right==null){ // case 1
	       if(N==root){
			   root = null;
		   }else{
			   P = findParent(N, root);
			   if(P.right==N) P.right=null;
			   else P.left = null;
		   }
	   }else if(N.right==null){ // case 2(left but no right child)
		   if(N==root){
			   root = N.right;
		   }else{
			   P = findParent(N, root);
			   if(P.right == N) P.right = N.left;
			   else P.left = N.left;
		   }
	   }else if(N.left==null){ // case 2(right but no left child)
		   if(N==root){
			   root = N.right;
		   }else{
			   P = findParent(N, root);
			   if(P.right==N) P.right = N.right;
			   else P.left = N.right;
		   }
		}else{                 // case 3 N.left!=null && N.right!=null
			S = findLeftmost(N.right);
			N.item.key = S.item.key;
			N.item.value = S.item.value;
			P = findParent(S, N);
			if(P.right==S) P.right = S.right;
			else P.left = S.right;
		}
		numPairs--;
   }
   
   // makeEmpty()
   // pre: none
   public void makeEmpty(){
	  deleteAll(root);
      root = null;
      numPairs = 0;
   }
   
   // toString()
   // returns a String representation of this Dictionary
   // overrides Object's toString() method
   // pre: none
   public String toString(){
      printInOrder(root);
	  return "";
   }
}
   
   

File: lab/lab5/Dictionary.c
//-----------------------------------------------------------------------------
// Dictionary.c
// Binary Search Tree implementation of the Dictionary ADT
//-----------------------------------------------------------------------------

#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<assert.h>
#include"Dictionary.h"

// private types and functions ------------------------------------------------

// NodeObj
typedef struct NodeObj{
   char* key;
   char* value;
   struct NodeObj* next;
} NodeObj;

// Node
typedef NodeObj* Node;

// newNode()
// constructor for private Node type
Node newNode(char* k, char* v) {
   Node N = malloc(sizeof(NodeObj));
   assert(N!=NULL);
   N->key = k;
   N->value = v;
   N->next = NULL;
   return(N);
}

// freeNode()
// destructor for private Node type
void freeNode(Node* pN){
   if( pN!=NULL && *pN!=NULL ){
      free(*pN);
      *pN = NULL;
   }
}

// DictionaryObj
typedef struct DictionaryObj{
   Node head;
   Node tail;
   int numItems;
} DictionaryObj;

// public functions -----------------------------------------------------------

// newDictionary()
// constructor for the Dictionary type
Dictionary newDictionary(){
   Dictionary D = malloc(sizeof(DictionaryObj));
   assert(D!=NULL);
   D->head = NULL;
   D->numItems = 0;
   return D;
}

// freeDictionary()
// destructor for the Dictionary type
void freeDictionary(Dictionary* pD){
   if( pD!=NULL && *pD!=NULL ){
      makeEmpty(*pD);
      free(*pD);
      *pD = NULL;
   }
}

// isEmpty()
// returns 1 (true) if D is empty, 0 (false) otherwise
// pre: none
int isEmpty(Dictionary D){
   if( D==NULL ){
      fprintf(stderr, 
         "Dictionary Error: calling isEmpty() on NULL Dictionary reference\n");
      exit(EXIT_FAILURE);
   }
   if(D->numItems>0){
	   return 0;
   }else{
	   return 1;
   }
}

// size()
// returns the number of (key, value) pairs in D
// pre: none
int size(Dictionary D){
   if( D==NULL ){
      fprintf(stderr, 
         "Dictionary Error: calling size() on NULL Dictionary reference\n");
      exit(EXIT_FAILURE);
   }
   return(D->numItems);
}

// lookup()
// returns the value v such that (k, v) is in D, or returns NULL if no 
// such value v exists.
// pre: none
char* lookup(Dictionary D, char* k){
   Node N = D->head;
   if( D==NULL ){
      fprintf(stderr, 
         "Dictionary Error: calling lookup() on NULL Dictionary reference\n");
      exit(EXIT_FAILURE);
   }
   while(N != NULL){
	   if(strcmp(N->key, k)==0)
              return N->value;
	      N = N->next;
   }
   return NULL;
}

// insert()
// inserts new (key,value) pair into D
// pre: lookup(D, k)==NULL
void insert(Dictionary D, char* k, char* v){
   if(lookup(D, k) != NULL){
	   fprintf(stderr, "Dictionary Error: cannot insert an existing key\n");
   }else if(D->head == NULL){
      Node N = newNode(k, v);
      D->head = N;
      D->numItems++;
   }else{
      Node N = D->head;
      while(N != NULL){
         if(N->next == NULL){
            break;
         }
         N = N->next;
      }
      N->next = newNode(k, v);
      D->numItems++; 
   }
}

// delete()
// deletes pair with the key k
// pre: lookup(D, k)!=NULL
void delete(Dictionary D, char* k){
   if( lookup(D, k)==NULL ){
      fprintf(stderr, 
         "Dictionary Error: cannot delete non-existent key\n");
      exit(EXIT_FAILURE);
   }else{
      if(D->numItems <= 1){
         Node N = D->head;
         D->head = (D->head)->next;
         N->next = NULL;
         D->numItems--;
      }else{
         Node N = D->head;
         if(strcmp((N->key), k)==0){
            D->head = N->next;
            D->numItems--;
         }else{
            while(!(strcmp((N->next->key),k)==0)){
               N = N->next;
            }
            N->next = N->next->next;
            D->numItems--;
         }
      }
   }
}

// makeEmpty()
// re-sets D to the empty state.
// pre: none
void makeEmpty(Dictionary D){
   D->head = NULL;
   D->tail = NULL;
   D->numItems = 0;
}

// printDictionary()
// pre: none
// prints a text representation of D to the file pointed to by out
void printDictionary(FILE* out, Dictionary D){
   Node N;
   if( D==NULL ){
      fprintf(stderr, 
         "Dictionary Error: calling printDictionary() on NULL"
         " Dictionary reference\n");
      exit(EXIT_FAILURE);
   }
   for(N=D->head; N!=NULL; N=N->next) fprintf(out, "%s %s \n", N->key, N->value);
   fprintf(out, "\n");
}

